src/pip/_internal/operations/install/wheel.py

```diff
@@ -583,6 +576,51 @@ def is_entrypoint_wrapper(name):
                 filter=filter,
             )
 
+    def pyc_source_file_paths():
+        # type: () -> Iterator[text_type]
+        # We de-duplicate installation paths, since there can be overlap (e.g.
+        # file in .data maps to same location as file in wheel root).
+        # Sorting installation paths makes it easier to reproduce and debug
+        # issues related to permissions on existing files.
+        for installed_path in sorted(set(installed.values())):
+            full_installed_path = os.path.join(lib_dir, installed_path)
+            if not os.path.isfile(full_installed_path):
+                continue
+            if not full_installed_path.endswith('.py'):
+                continue
+            yield full_installed_path
+
+    def pyc_output_path(path):
+        # type: (text_type) -> text_type
+        """Return the path the pyc file would have been written to.
+        """
+        if PY2:
+            if sys.flags.optimize:
+                return path + 'o'
+            else:
+                return path + 'c'
+        else:
+            return importlib.util.cache_from_source(path)
+
+    # Compile all of the pyc files for the installed files
+    if pycompile:
+        with captured_stdout() as stdout:
+            with warnings.catch_warnings():
+                warnings.filterwarnings('ignore')
+                for path in pyc_source_file_paths():
+                    # Python 2's `compileall.compile_file` is compatible with
+                    # unicode input, but the declared type doesn't indicate
+                    # it.
+                    path_arg = cast('str', path)
+                    success = compileall.compile_file(
```

Comment:
Using `py_compile.compile` would be better, as it returns the path of the compiled file, so we don't have to rely on recalculating it ourselves. The `force` and `quiet` arguments don't exist for that function, though, so we would need to change the logic somehow. I'm fine with doing that in a follow-up PR (and it's not a priority).

Reply:
Agreed. We'd still need to manually get the path on Python 2, unfortunately, but it should be about the same amount of code. I had originally gone that route, but figured the jump from `compileall.compile_dir` to `compileall.compile_file` was big enough, and would at least help get us closer.

Reply:
3 more months. Just 3 more.